README.md
&nbsp;

<div align="center">

<img src="https://raw.githubusercontent.com/AgnostiqHQ/covalent-braket-plugin/main/assets/aws_braket_readme_banner.jpg" width=150%>

[![covalent](https://img.shields.io/badge/covalent-0.177.0-purple)](https://github.com/AgnostiqHQ/covalent)
[![python](https://img.shields.io/pypi/pyversions/covalent-braket-plugin)](https://github.com/AgnostiqHQ/covalent-braket-plugin)
[![tests](https://github.com/AgnostiqHQ/covalent-braket-plugin/actions/workflows/tests.yml/badge.svg)](https://github.com/AgnostiqHQ/covalent-braket-plugin/actions/workflows/tests.yml)
[![codecov](https://codecov.io/gh/AgnostiqHQ/covalent-braket-plugin/branch/main/graph/badge.svg?token=QNTR18SR5H)](https://codecov.io/gh/AgnostiqHQ/covalent-braket-plugin)
[![apache](https://img.shields.io/badge/License-Apache_License_2.0-blue)](https://www.apache.org/licenses/LICENSE-2.0)

</div>

## Covalent Braket Hybrid Jobs Plugin

Covalent is a Pythonic workflow tool used to execute tasks on advanced computing hardware. This executor plugin interfaces Covalent with [AWS Braket Hybrid Jobs](https://docs.aws.amazon.com/braket/latest/developerguide/braket-jobs.html)
## Installing

To use this plugin with Covalent, install it with `pip`:
```
pip install covalent-braket-plugin
```

## Usage Example

The following workflow prepares a uniform superposition of the single-qubit standard basis states and measures it.

```python
import covalent as ct
from covalent_braket_plugin.braket import BraketExecutor
import os

# AWS resources to pass to the executor
credentials_file = "~/.aws/credentials"
profile = "default"
s3_bucket_name = "braket_s3_bucket"
ecr_image_uri = "111223344.dkr.ecr.us-east-1.amazonaws.com/amazon-braket-ecr-repo:latest"
iam_role_name = "covalent-braket-iam-role"

ex = BraketExecutor(
    profile=profile,
    credentials=credentials_file,
    s3_bucket_name=s3_bucket_name,
    ecr_image_uri=ecr_image_uri,
    braket_job_execution_role_name=iam_role_name,
    quantum_device="arn:aws:braket:::device/quantum-simulator/amazon/sv1",
    classical_device="ml.m5.large",
    storage=30,
    time_limit=300,
)


@ct.electron(executor=ex)
def simple_quantum_task(num_qubits: int):
    import pennylane as qml

    # These are passed to the Hybrid Jobs container at runtime
    device_arn = os.environ["AMZN_BRAKET_DEVICE_ARN"]
    s3_bucket = os.environ["AMZN_BRAKET_OUT_S3_BUCKET"]
    s3_task_dir = os.environ["AMZN_BRAKET_TASK_RESULTS_S3_URI"].split(s3_bucket)[1]

    device = qml.device(
        "braket.aws.qubit",
        device_arn=device_arn,
        s3_destination_folder=(s3_bucket, s3_task_dir),
        wires=num_qubits,
    )

    @qml.qnode(device=device)
    def simple_circuit():
        qml.Hadamard(wires=[0])
        return qml.expval(qml.PauliZ(wires=[0]))

    res = simple_circuit().numpy()
    return res


@ct.lattice
def simple_quantum_workflow(num_qubits: int):
    return simple_quantum_task(num_qubits=num_qubits)


dispatch_id = ct.dispatch(simple_quantum_workflow)(1)
result_object = ct.get_result(dispatch_id, wait=True)

# We expect 0 as the result
print("Result:", result_object.result)
```

To run such workflows, users must have AWS credentials allowing access
to Braket, ECR, S3, and some other services. These permissions must be
defined in an IAM Role (called `"covalent-braket-iam-role"` in this
example). The [AWS
documentation has more information about managing Braket
access](https://docs.aws.amazon.com/braket/latest/developerguide/braket-manage-access.html).

## Overview of Configuration

See the
[RTD](https://covalent.readthedocs.io/en/latest/api/executors/awsbraket.html)
for how to configure this executor.

## Required Cloud Resources

In order to run your workflows with covalent there are a few notable resources that need to be provisioned first. Particularly an S3 bucket must be created, an IAM role with the `AmazonBraketFullAccess` policy, and a private ECR repo with an uploaded image for the tasks to use.

For more information regarding which cloud resources need to be provisioned visit our read the docs [RTD](https://covalent.readthedocs.io/en/latest/api/executors/awsbraket.html) guide for this plugin.

## Release Notes

Release notes are available in the [Changelog](https://github.com/AgnostiqHQ/covalent-braket-plugin/blob/main/CHANGELOG.md).

## Citation

Please use the following citation in any publications:

> W. J. Cunningham, S. K. Radha, F. Hasan, J. Kanem, S. W. Neagle, and S. Sanand.
> *Covalent.* Zenodo, 2022. https://doi.org/10.5281/zenodo.5903364

## License

Covalent is licensed under the Apache 2.0 License. See the
[LICENSE](https://github.com/AgnostiqHQ/covalent-braket-plugin/blob/main/LICENSE)
file or contact the [support team](mailto:support@agnostiq.ai) for
more details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [UNRELEASED]

## [0.28.0] - 2023-11-03

### Added

- Added terraform files to the repo for deploying necessary resources

### Changed

- Removed the `--use-feature=in-tree-build` flag from the `pip install` command in the Dockerfile as it was causing the build to fail

## [0.27.0] - 2023-09-20

### Changed

- Updated license to Apache

## [0.26.0] - 2023-06-27

### Added

- Add docstring to __init__.

### Changed

- Updates __init__ signature kwargs replaced with parent for better documentation.

## [0.25.0] - 2023-06-20

### Changed

- Implement braket executor cancel task

## [0.24.1] - 2023-04-26

### Fixed

- Fixed `docker` workflow referring to deprecated `stable_version` input

## [0.24.0] - 2023-04-26

### Changed

- Added configurable covalent version in `docker` workflow to build base executor image

## [0.23.0] - 2023-04-26

### Added

- Print statement to functional tests.

## [0.22.2] - 2023-02-13

### Fixed

- Minor fix to tests workflow.

## [0.22.1] - 2023-02-10

### Fixed

- Git config safe repo flag now points to the correct path environment variable.

### Operations

-  Fix expression in docker yml.

## [0.22.0] - 2023-02-09

### Added

- Adding `repository_dispatch` trigger to build base executor image from core covalent

## [0.21.2] - 2023-02-09

### Fixed

- Second attempt at fixing tests.yml workflow.

## [0.21.1] - 2023-02-07

### Fixed

- tests.yml workflow.

## [0.21.0] - 2023-01-19

### Changed

- Updated docker workflow to allow manual pre-release base executor image release
- Minor updates for consistency with other docker workflows

## [0.20.1] - 2023-01-18

### Fixed

- Fixed Dockerfile and docker workflow syntax to build pre-release base executor images

## [0.20.0] - 2023-01-07

### Changed

- Updated docker workflow to specify pre-release flag as a build arg

## [0.19.0] - 2023-01-06

### Changed

- Ability to specify covalent version in docker workflow

## [0.18.3] - 2023-01-06

### Fixed

- Fixed COVALENT_BASE_IMAGE arg in docker workflow quotations

## [0.18.2] - 2023-01-06

### Fixed

- Fixed COVALENT_BASE_IMAGE arg in docker workflow

## [0.18.1] - 2023-01-06

### Fixed

- Fixed docker workflow github event name syntax

## [0.18.0] - 2023-01-06

### Changed

- Removed inputs.version from checkout step in docker workflow

## [0.17.3] - 2023-01-06

### Fixed

- Added id-token write, and permissions for OIDC to push to ECR from role

## [0.17.2] - 2023-01-06

### Fixed

- Fixed `typo` in action spelling

## [0.17.1] - 2023-01-06

### Fixed

- Updated version of the `docker` buildx actions

## [0.17.0] - 2023-01-06

### Changed

- Updated docker workflow to create manual pre-releases in addition to stable version

## [0.16.0] - 2023-01-06

### Operations

### Added

- Adding `docker.yml` workflow to build and release braket executor docker image

### Changed

- Removed explicit `covalent` installation in the Docker and use base `covalent` image from public registry
- Invoke `docker.yml` from `release.yml` via a workflow call

## [0.15.0] - 2022-12-15

### Changed

- Removed references to `.env` file in the functional test README.

## [0.14.0] - 2022-12-06

### Changed

- Using executor aliases instead of classes for functional tests

## [0.13.0] - 2022-11-22

### Changed

- Removed executor defaults for boto3 session args, added .env.example

## [0.12.0] - 2022-11-10

### Changed

- Functional tests using pytest and .env file configuration

## [0.11.0] - 2022-10-28

### Changed

- Bumped aws plugins version to new stable release

## [0.10.0] - 2022-10-27

### Changed

- Added Alejandro to paul blart group

## [0.9.0] - 2022-10-27


### Changed

- Updated Dockerfile to contain `COVALENT_BASE_IMAGE` ARG
- Removed explicit cloudpickle install due to version conflict with covalent cloudpickle from Dockerfile
- Constrained covalent semver range to be less than major version 1 for this version of Dockerfile

### Documentation

- Update README to include reference to new `ecr_image_uri` key


## [0.8.0] - 2022-10-25

### Changed

- Updated executor to use pre-built braket executor base image
- Updated aws plugins to be gt than 0.7.0rc0

## [0.7.2] - 2022-10-06

### Fixed

- License checker reference.

### Operations

- Added license workflow

## [0.7.1] - 2022-10-06

### Fixed

- Store `BASE_COVALENT_AWS_PLUGINS_ONLY` in a temporary file rather than storing it as an environment variable.

### Docs

- Simplified README

## [0.7.0] - 2022-09-30

### Added

-  Logic to specify that only the base covalent-aws-plugins package is to be installed.

## [0.6.1] - 2022-09-20

### Fixed

- Getting default values from config file if not defined in class instance

## [0.6.0] - 2022-09-16

### Changed

- Updated requirements.txt to pin aws executor plugins to pre-release version 0.1.0rc0

## [0.5.0] - 2022-09-15

### Changed

- BraketExecutor is now derived from AWSExecutor

### Tests

- Enabled Codecov
- Added tests
- Update pre-commit hooks
- Updated some tests to be async aware

## [0.4.1] - 2022-08-23

### Fixed

- Removed `--pre` from `pip install covalent` in Dockerfile

## [0.4.0] - 2022-08-17

### Changed

- Fixed `covalent` version to `stable` release

## [0.3.0] - 2022-08-16

### Added

- Updated required `covalent` version

## [0.2.0] - 2022-08-13

### Added

- Workflows needed for release

## [0.1.2] - 2022-08-10

### Fixed

- Dynamically set bucket name and repo name

### Docs

- Updated README with more getting started details

## [0.1.1] - 2022-08-08

### Fixed

- Dockerfile so that SAGEMAKER_PROGRAM is set correctly, removing ENTRYPOINT and CMD
- Pennylane version, upgraded to 0.24.0

## [0.1.0] - 2022-03-31

### Changed


test.py
import covalent as ct
from typing import List

#@ct.electron(executor="local")
@ct.electron(executor="braket")
def hybrid_task(size: int, shots: int, angles: List):
    import pennylane as qml
    import random
    import os
    device_arn = os.environ["AMZN_BRAKET_DEVICE_ARN"]
    s3_bucket = os.environ["AMZN_BRAKET_OUT_S3_BUCKET"]
    s3_task_dir = os.environ["AMZN_BRAKET_TASK_RESULTS_S3_URI"].split(s3_bucket)[1]

#    os.environ["AWS_SHARED_CREDENTIALS_FILE"] = "/Users/scott/.aws/credentials"
#    os.environ["AWS_PROFILE"] = "SDLC-Administrator"
#    device_arn = "arn:aws:braket:::device/quantum-simulator/amazon/sv1"
#    s3_bucket = "amazon-braket-a0d10035b91d"
#    s3_task_dir = "covalent-braket-plugin"

    device = qml.device(
        "braket.aws.qubit", 
        device_arn=device_arn, 
        s3_destination_folder=(s3_bucket, s3_task_dir), 
        wires=size,
        shots=shots,
        parallel=True,
        max_parallel=4,
    )

    @qml.qnode(device)
    def circuit(angles):
        for angle in angles:
            wire = random.randint(0, size - 1)
            if random.random() < 0.5:
                qml.RX(angle, wires=wire)
            else:
                qml.RY(angle, wires=wire)

        return [qml.expval(qml.PauliZ(i)) for i in range(size)]

    result = circuit(angles=angles)

    return result

@ct.lattice
def workflow(size: int, shots: int, angles: List):
    return hybrid_task(size, shots, angles)

if __name__ == "__main__":
#    result = hybrid_task(size=2, shots=100, angles=[0.2, 0.3, 0.4, 0.5])
#    print(result)
    result = ct.dispatch_sync(workflow)(size=2, shots=100, angles=[0.2, 0.3, 0.4, 0.5])
    print(result)


tests/test.py
from typing import List

import covalent as ct


@ct.electron(executor="braket")
def hybrid_task(size: int, shots: int, angles: List):
    import os
    import random

    import pennylane as qml

    device_arn = os.environ["AMZN_BRAKET_DEVICE_ARN"]
    s3_bucket = os.environ["AMZN_BRAKET_OUT_S3_BUCKET"]
    s3_task_dir = os.environ["AMZN_BRAKET_TASK_RESULTS_S3_URI"].split(s3_bucket)[1]

    device = qml.device(
        "braket.aws.qubit",
        device_arn=device_arn,
        s3_destination_folder=(s3_bucket, s3_task_dir),
        wires=size,
        shots=shots,
        parallel=True,
        max_parallel=4,
    )

    @qml.qnode(device)
    def circuit(angles):
        for angle in angles:
            wire = random.randint(0, size - 1)
            if random.random() < 0.5:
                qml.RX(angle, wires=wire)
            else:
                qml.RY(angle, wires=wire)

        return [qml.expval(qml.PauliZ(i)) for i in range(size)]

    result = circuit(angles=angles)

    return result

@ct.lattice
def workflow(size: int, shots: int, angles: List):
    return hybrid_task(size, shots, angles)

if __name__ == "__main__":
    result = ct.dispatch_sync(workflow)(size=2, shots=100, angles=[0.2, 0.3, 0.4, 0.5])
    print(result)


tests/functional_test.py
from typing import List

import covalent as ct
from covalent._results_manager import Result


def test_executor_functional():
    @ct.electron(executor="braket")
    def hybrid_task(size: int, shots: int, angles: List):
        import os
        import random

        import pennylane as qml

        device_arn = os.environ["AMZN_BRAKET_DEVICE_ARN"]
        s3_bucket = os.environ["AMZN_BRAKET_OUT_S3_BUCKET"]
        s3_task_dir = os.environ["AMZN_BRAKET_TASK_RESULTS_S3_URI"].split(s3_bucket)[1]

        device = qml.device(
            "braket.aws.qubit",
            device_arn=device_arn,
            s3_destination_folder=(s3_bucket, s3_task_dir),
            wires=size,
            shots=shots,
            parallel=True,
            max_parallel=4,
        )

        @qml.qnode(device)
        def circuit(angles):
            for angle in angles:
                wire = random.randint(0, size - 1)
                if random.random() < 0.5:
                    qml.RX(angle, wires=wire)
                else:
                    qml.RY(angle, wires=wire)

            return [qml.expval(qml.PauliZ(i)) for i in range(size)]

        result = circuit(angles=angles)

        return result

    @ct.lattice
    def workflow(size: int, shots: int, angles: List):
        return hybrid_task(size, shots, angles)

    result = ct.dispatch_sync(workflow)(size=2, shots=100, angles=[0.2, 0.3, 0.4, 0.5])
    assert result.status == str(Result.COMPLETED)


covalent_braket_plugin/exec.py
import os

import boto3
import cloudpickle as pickle

s3_bucket_name = os.environ.get("SM_HP_S3_BUCKET_NAME")
result_filename = os.environ.get("SM_HP_RESULT_FILENAME")
func_filename = os.environ.get("SM_HP_COVALENT_TASK_FUNC_FILENAME")
work_dir = os.environ.get("SM_HP_WORKDIR", "/opt/ml/code")

print(f"Covalent artifact s3 bucket: {s3_bucket_name}")
print(f"Result filename: {result_filename}")
print(f"Function filename: {func_filename}")

local_func_filename = os.path.join(work_dir, func_filename)
local_result_filename = os.path.join(work_dir, result_filename)

s3 = boto3.client("s3")
s3.download_file(s3_bucket_name, func_filename, local_func_filename)

with open(local_func_filename, "rb") as f:
    function, args, kwargs = pickle.load(f)

result = function(*args, **kwargs)

with open(local_result_filename, "wb") as f:
    pickle.dump(result, f)

s3.upload_file(local_result_filename, s3_bucket_name, result_filename)


covalent_braket_plugin/scripts.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the GNU Affero General Public License 3.0 (the "License").
# A copy of the License may be obtained with this software package or at
#
#      https://www.gnu.org/licenses/agpl-3.0.en.html
#
# Use of this file is prohibited except in compliance with the License. Any
# modifications or derivative works of this file must retain this copyright
# notice, and modified files must contain a notice indicating that they have
# been altered from the originals.
#
# Covalent is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE. See the License for more details.
#
# Relief from the License may be granted by purchasing a commercial license.

"""Docker and Python execution scripts."""

PYTHON_EXEC_SCRIPT = """
import os
import boto3
import cloudpickle as pickle

local_func_filename = os.path.join("{docker_working_dir}", "{func_filename}")
local_result_filename = os.path.join("{docker_working_dir}", "{result_filename}")

s3 = boto3.client("s3")
s3.download_file("{s3_bucket_name}", "{func_filename}", local_func_filename)

with open(local_func_filename, "rb") as f:
    function, args, kwargs = pickle.load(f)

result = function(*args, **kwargs)

with open(local_result_filename, "wb") as f:
    pickle.dump(result, f)

s3.upload_file(local_result_filename, "{s3_bucket_name}", "{result_filename}")
"""

DOCKER_SCRIPT = """
FROM amd64/python:3.8-slim-buster

RUN apt-get update && apt-get install -y \\
  gcc \\
  && rm -rf /var/lib/apt/lists/*
RUN pip install --no-cache-dir --use-feature=in-tree-build --upgrade \\
  amazon-braket-pennylane-plugin==1.6.9 \\
  boto3==1.20.48 \\
  cloudpickle==2.0.0 \\
  pennylane==0.24.0 \\
  sagemaker-training
RUN pip install covalent

WORKDIR {docker_working_dir}

COPY {func_basename} {docker_working_dir}
ENV SAGEMAKER_PROGRAM {func_basename}
"""


covalent_braket_plugin/braket_container_script.py
import json
import os
import pickle
from pathlib import Path

import boto3

# "covalent-{dispatch_id}-{node_id}"
job_name = os.environ.get("AMZN_BRAKET_JOB_NAME")

# {dispatch_id}-{node_id}
task_id = job_name[len("covalent-") :]

working_dir = "/opt/ml/code/workdir"
Path(working_dir).mkdir(parents=True, exist_ok=True)

meta_dir = os.environ.get("SM_CHANNEL_TASKMETADATA")
meta_file_path = os.path.join(meta_dir, f"metadata-{task_id}.json")

with open(meta_file_path, "r") as f:
    metadata = json.load(f)

input_object_key = metadata["input_object_key"]
output_object_key = metadata["output_object_key"]
s3_bucket = metadata["s3_bucket"]


FUNC_FILENAME = f"func-{task_id}.pkl"
RESULT_FILENAME = f"result-{task_id}.json"

local_func_filename = os.path.join(working_dir, FUNC_FILENAME)
local_result_filename = os.path.join(working_dir, RESULT_FILENAME)

s3 = boto3.client("s3")
s3.download_file(s3_bucket, input_object_key, local_func_filename)

with open(local_func_filename, "rb") as f:
    function, args, kwargs = pickle.load(f)

# Result is a TransportableObject
result = function(*args, **kwargs)

with open(local_result_filename, "w") as f:
    f.write(result.serialize_to_json())

s3.upload_file(local_result_filename, s3_bucket, output_object_key)
print(f"Uploaded result to s3://{s3_bucket}/{output_object_key}")


covalent_braket_plugin/braket.py
# Copyright 2021 Agnostiq Inc.
#
# This file is part of Covalent.
#
# Licensed under the Apache License 2.0 (the "License"). A copy of the
# License may be obtained with this software package or at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Use of this file is prohibited except in compliance with the License.
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""AWS Braket Hybrid Jobs executor plugin for the Covalent dispatcher."""

import asyncio
import os
import sys
import tempfile
from functools import partial
from pathlib import Path
from typing import Any, Callable, Dict, List

import boto3
import botocore
import cloudpickle as pickle
from covalent._shared_files.config import get_config
from covalent._shared_files.exceptions import TaskCancelledError
from covalent._shared_files.logger import app_log
from covalent_aws_plugins import AWSExecutor

_EXECUTOR_PLUGIN_DEFAULTS = {
    "credentials": "~/.aws/credentials",
    "profile": "default",
    "region": "us-east-1",
    "s3_bucket_name": os.environ.get("BRAKET_COVALENT_S3")
    or "amazon-braket-covalent-job-resources",
    "ecr_image_uri": "",
    "braket_job_execution_role_name": "CovalentBraketJobsExecutionRole",
    "quantum_device": "arn:aws:braket:::device/quantum-simulator/amazon/sv1",
    "classical_device": "ml.m5.large",
    "storage": 30,
    "time_limit": 300,
    "cache_dir": "/tmp/covalent",
    "poll_freq": 10,
}
BRAKET_JOB_NAME = "job-{dispatch_id}-{node_id}"
executor_plugin_name = "BraketExecutor"


class BraketExecutor(AWSExecutor):
    """AWS Braket Hybrid Jobs executor plugin class."""

    def __init__(
        self,
        ecr_image_uri: str = None,
        s3_bucket_name: str = None,
        braket_job_execution_role_name: str = None,
        classical_device: str = None,
        storage: int = None,
        time_limit: int = None,
        poll_freq: int = None,
        quantum_device: str = None,
        profile: str = None,
        credentials: str = None,
        cache_dir: str = None,
        region: str = None,
        log_group_name: str = None,
    ):
        """
        Initialize the Braket executor plugin.

        Args:
            ecr_image_uri (str): The URI of the ECR image to use for the Braket jobs.
            s3_bucket_name (str): The name of the S3 bucket to use for the Braket jobs.
            braket_job_execution_role_name (str): The name of the IAM role to use for the Braket jobs.
            classical_device (str): The name of the classical device to use for the Braket jobs.
            storage (int): The amount of storage to use for the Braket jobs.
            time_limit (int): The time limit for the Braket jobs.
            poll_freq (int): The polling frequency for the Braket jobs.
            quantum_device (str): The name of the quantum device to use for the Braket jobs.
            profile (str): The name of the AWS profile to use for the Braket jobs.
            credentials (str): The path to the AWS credentials file to use for the Braket jobs.
            cache_dir (str): The path to the cache directory to use for the Braket jobs.
            region (str): The name of the AWS region to use for the Braket jobs.
            log_group_name (str): The name of the CloudWatch log group to use for the Braket jobs.
        """

        region = region or get_config("executors.braket.region")
        credentials = credentials or get_config("executors.braket.credentials")
        profile = profile or get_config("executors.braket.profile")
        s3_bucket_name = s3_bucket_name or get_config("executors.braket.s3_bucket_name")
        braket_job_execution_role_name = braket_job_execution_role_name or get_config(
            "executors.braket.braket_job_execution_role_name"
        )
        s3_bucket_name = s3_bucket_name or get_config("executors.braket.s3_bucket_name")
        cache_dir = cache_dir or get_config("executors.braket.cache_dir")
        time_limit = time_limit or get_config("executors.braket.time_limit")
        poll_freq = poll_freq or get_config("executors.braket.poll_freq")

        super().__init__(
            credentials_file=credentials,
            profile=profile,
            s3_bucket_name=s3_bucket_name,
            execution_role=braket_job_execution_role_name,
            region=region,
            cache_dir=cache_dir,
            time_limit=time_limit,
            poll_freq=poll_freq,
            log_group_name=log_group_name,
        )

        self.quantum_device = quantum_device or get_config("executors.braket.quantum_device")
        self.classical_device = classical_device or get_config("executors.braket.classical_device")
        self.storage = storage or get_config("executors.braket.storage")
        self.ecr_image_uri = ecr_image_uri or get_config("executors.braket.ecr_image_uri")

    async def _execute_partial_in_threadpool(self, partial_func):
        loop = asyncio.get_running_loop()
        future = await loop.run_in_executor(None, partial_func)
        return future

    def load_pickle(self, filename, remove_file):
        with open(filename, "rb") as f:
            result = pickle.load(f)
        if remove_file:
            os.remove(filename)
        return result

    async def _upload_task(
        self, function: Callable, args: List, kwargs: Dict, upload_metadata: Dict
    ):
        """
        Abstract method that uploads the pickled function to the remote cache.
        """
        image_tag = upload_metadata["image_tag"]
        func_filename = f"func-{image_tag}.pkl"

        s3 = boto3.Session(**self.boto_session_options()).client("s3")

        with tempfile.NamedTemporaryFile(dir=self.cache_dir) as function_file:
            # Write serialized function to file
            pickle.dump((function, args, kwargs), function_file)
            function_file.flush()

            # Upload pickled function to S3
            await self._execute_partial_in_threadpool(
                partial(s3.upload_file, function_file.name, self.s3_bucket_name, func_filename)
            )

    async def submit_task(self, submit_metadata: Dict) -> Any:
        """
        Abstract method that invokes the task on the remote backend.

        Args:
            task_metadata: Dictionary of metadata for the task. Current keys are
                          `dispatch_id` and `node_id`.
        Return:
            task_uuid: Task UUID defined on the remote backend.
        """
        braket = boto3.Session(**self.boto_session_options()).client("braket")

        image_tag = submit_metadata["image_tag"]
        result_filename = submit_metadata["result_filename"]
        account = submit_metadata["account"]

        func_filename = f"func-{image_tag}.pkl"

        app_log.debug(f"Using ECR Image URI: {self.ecr_image_uri}")
        args = {
            "hyperParameters": {
                "COVALENT_TASK_FUNC_FILENAME": func_filename,
                "RESULT_FILENAME": result_filename,
                "S3_BUCKET_NAME": self.s3_bucket_name,
            },
            "algorithmSpecification": {
                "containerImage": {
                    "uri": self.ecr_image_uri,
                },
            },
            "checkpointConfig": {
                "s3Uri": f"s3://{self.s3_bucket_name}/checkpoints/{image_tag}",
            },
            "deviceConfig": {
                "device": self.quantum_device,
            },
            "instanceConfig": {
                "instanceType": self.classical_device,
                "volumeSizeInGb": self.storage,
            },
            "jobName": f"covalent-{image_tag}",
            "outputDataConfig": {
                "s3Path": f"s3://{self.s3_bucket_name}/braket/{image_tag}",
            },
            "roleArn": f"arn:aws:iam::{account}:role/{self.execution_role}",
            "stoppingCondition": {
                "maxRuntimeInSeconds": self.time_limit,
            },
        }

        try:
            app_log.debug("Submitting Braket Job:")
            app_log.debug(args)

            job = await self._execute_partial_in_threadpool(partial(braket.create_job, **args))

        except botocore.exceptions.ClientError as error:
            app_log.debug(error.response)
            raise error

        return job["jobArn"]

    async def _poll_task(self, poll_metadata: Dict) -> Any:
        """
        Abstract method that polls the remote backend until the status of a workflow's execution
        is either COMPLETED or FAILED.
        """
        braket = boto3.Session(**self.boto_session_options()).client("braket")
        job_arn = poll_metadata["job_arn"]

        status = await self.get_status(braket, job_arn)

        while status not in ["COMPLETED", "FAILED", "CANCELLED"]:
            await asyncio.sleep(self.poll_freq)
            status = await self.get_status(braket, job_arn)

        if status == "FAILED":
            job = await self._execute_partial_in_threadpool(
                partial(braket.get_job, jobArn=job_arn)
            )
            failure_reason = job["failureReason"]
            raise Exception(failure_reason)

    async def query_result(self, query_metadata: Dict) -> Any:
        """
        Abstract method that retrieves the pickled result from the remote cache.
        """

        s3 = boto3.Session(**self.boto_session_options()).client("s3")
        logs = boto3.Session(**self.boto_session_options()).client("logs")

        result_filename = query_metadata["result_filename"]
        task_results_dir = query_metadata["task_results_dir"]
        image_tag = query_metadata["image_tag"]

        local_result_filename = os.path.join(task_results_dir, result_filename)

        await self._execute_partial_in_threadpool(
            partial(s3.download_file, self.s3_bucket_name, result_filename, local_result_filename)
        )

        result = await self._execute_partial_in_threadpool(
            partial(self.load_pickle, local_result_filename, True)
        )

        log_group_name = "/aws/braket/jobs"
        log_stream_prefix = f"covalent-{image_tag}"

        log_stream_describe = await self._execute_partial_in_threadpool(
            partial(
                logs.describe_log_streams,
                logGroupName=log_group_name,
                logStreamNamePrefix=log_stream_prefix,
            )
        )
        log_stream_name = log_stream_describe["logStreams"][0]["logStreamName"]

        # TODO: This should be paginated, but the command doesn't support boto3 pagination
        # Up to 10000 log events can be returned from a single call to get_log_events()
        all_log_events = await self._execute_partial_in_threadpool(
            partial(
                logs.get_log_events,
                logGroupName=log_group_name,
                logStreamName=log_stream_name,
            )
        )
        events = all_log_events["events"]

        log_events = "".join(event["message"] + "\n" for event in events)

        # output, stdout, stderr
        return result, log_events, ""

    async def cancel(self, task_metadata: Dict, job_handle: str) -> bool:
        """
        Cancel the quantum task

        Args:
            task_metadata: Dictionary with the task's dispatch_id and node id.
            job_handle: Unique job ARN assigned to the task by AWS Braket.

        Returns:
            If the job was cancelled or not
        """
        try:
            braket = boto3.Session(**self.boto_session_options()).client("braket")
            partial_func = partial(braket.cancel_quantum_task, quantumTaskArn=job_handle)
            await self._execute_partial_in_threadpool(partial_func)
            return True
        except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as error:
            app_log.debug(
                f"Failed to cancel Braket quantum task with task metadata: \
                          {task_metadata} and error:{error}"
            )
            return False

    async def run(self, function: Callable, args: List, kwargs: Dict, task_metadata: Dict):

        dispatch_id = task_metadata["dispatch_id"]
        node_id = task_metadata["node_id"]
        results_dir = task_metadata["results_dir"]

        result_filename = f"result-{dispatch_id}-{node_id}.pkl"
        task_results_dir = os.path.join(results_dir, dispatch_id)
        image_tag = f"{dispatch_id}-{node_id}"
        batch_job_name = BRAKET_JOB_NAME.format(dispatch_id=dispatch_id, node_id=node_id)

        app_log.debug("Validating credentials...")
        # AWS Account Retrieval
        identity = await self._execute_partial_in_threadpool(
            partial(self._validate_credentials, raise_exception=True)
        )
        account = identity.get("Account")

        # TODO: Move this to BaseExecutor
        Path(self.cache_dir).mkdir(parents=True, exist_ok=True)

        upload_task_metadata = {
            "image_tag": image_tag,
        }

        if await self.get_cancel_requested():
            raise TaskCancelledError(f"Batch job {batch_job_name} requested to be cancelled")
        await self._upload_task(function, args, kwargs, upload_task_metadata)

        submit_metadata = {
            "image_tag": image_tag,
            "account": account,
            "task_results_dir": task_results_dir,
            "result_filename": result_filename,
        }

        app_log.debug("Submit metadata:")
        app_log.debug(submit_metadata)

        if await self.get_cancel_requested():
            raise TaskCancelledError(f"Batch job {batch_job_name} requested to be cancelled")
        job_arn = await self.submit_task(submit_metadata)

        poll_metadata = {"job_arn": job_arn}

        await self.set_job_handle(handle=job_arn)

        await self._poll_task(poll_metadata)

        query_metadata = {
            "result_filename": result_filename,
            "task_results_dir": task_results_dir,
            "job_arn": job_arn,
            "image_tag": image_tag,
        }

        output, stdout, stderr = await self.query_result(query_metadata)

        print(stdout, end="", file=sys.stdout)
        print(stderr, end="", file=sys.stderr)

        return output

    async def get_status(self, braket, job_arn: str) -> str:
        """Query the status of a previously submitted Braket hybrid job.

        Args:
            braket: Braket client object.
            job_arn: ARN used to identify a Braket hybrid job.

        Returns:
            status: String describing the job status.
        """
        app_log.debug(f"Getting Braket Job {job_arn} status...")
        job = await self._execute_partial_in_threadpool(partial(braket.get_job, jobArn=job_arn))
        status = job["status"]
        app_log.debug(f"Braket Job {job_arn} Status: {status}")
        return status
